apps/backend/src/application-period/application-period.service.ts

```diff
@@ -0,0 +1,110 @@
+import { BadRequestException, Injectable } from '@nestjs/common';
+import { ApplicationPeriod } from '@prisma/client';
+import { PrismaService } from 'nestjs-prisma';
+
+import { CreateApplicationPeriodDto } from './dto/create-application-period.dto';
+import { GetApplicationPeriodsDto } from './dto/get-application-periods.dto';
+import { UpdateApplicationPeriodDto } from './dto/update-application-period.dto';
+
+@Injectable()
+export class ApplicationPeriodService {
+  constructor(private readonly prisma: PrismaService) {}
+  findAll(getApplicationPeriodsDto: GetApplicationPeriodsDto) {
+    const skip = getApplicationPeriodsDto.page * getApplicationPeriodsDto.page_size;
+    return this.prisma.applicationPeriod.findMany({
+      skip,
+      take: Number(getApplicationPeriodsDto.page_size),
+    });
+  }
+  async getCurrentPeriod(): Promise<ApplicationPeriod> {
+    const period = await this.prisma.applicationPeriod.findFirst({
+      where: {
+        applicationPeriodStartAt: {
+          lte: new Date(),
+        },
+        applicationPeriodEndAt: {
+          gte: new Date(),
+        },
+      },
+    });
+    if (!period) {
+      throw new Error('No current application period found');
+    }
+    return period;
+  }
+
+  async findOne(id: number): Promise<ApplicationPeriod> {
+    const period = await this.prisma.applicationPeriod.findUnique({
+      where: {
+        id,
+      },
+    });
+    if (!period) {
+      throw new BadRequestException('Application period not found');
+    }
+    return period;
+  }
+
+  async create(createApplicationPeriodDto: CreateApplicationPeriodDto, user): Promise<ApplicationPeriod> {
+    const periods = await this.prisma.applicationPeriod.findMany();
+    const newPeriodStart = new Date(createApplicationPeriodDto.applicationPeriodStartAt);
+    const newPeriodEnd = new Date(createApplicationPeriodDto.applicationPeriodEndAt);
+    periods.forEach((period) => {
+      const start = new Date(period.applicationPeriodStartAt);
+      const end = new Date(period.applicationPeriodEndAt);
+      if (start <= newPeriodEnd && end >= newPeriodStart) {
+        throw new BadRequestException('Application period overlaps with another period');
+      }
+    });
```

Comment:
possibly better solution: https://github.com/Tschonti/pontozo/blob/master/packages/functions/src/functions/seasons/create.ts#L21-L30

(elements in the where array are connected with the OR operator)

Reply:
wow yours is a much better solution xd